Fixing tolerations in the k8s manifests

heschlie opened the pull request:
## Description

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->
Updating the k8s Calico manifests to use the newer tolerations so calico-node pods can be scheduled on the master node when tainted.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
None required
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/projectcalico/calico?pullRequest=1634) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/projectcalico/calico?pullRequest=1634) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/projectcalico/calico?pullRequest=1634) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/projectcalico/calico?pullRequest=1634) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/projectcalico/calico?pullRequest=1634) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/projectcalico/calico?pullRequest=1634) it.</sub>

caseydavenport commented on master/getting-started/kubernetes/installation/hosted/calico.yaml:
```diff
@@ -106,10 +106,15 @@ spec:
-          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
-           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
     spec:
+       tolerations:
```
For `calico/node`, can I suggest we use the same tolerations as are used in GKE?

https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/calico-policy-controller/calico-node-daemonset.yaml#L151-L158

This lets calico/node run on ALL nodes in the cluster, no matter the toleration. I think this is desirable in 99% of cases.

heschlie replied:
Seems like a good idea, I'll add the missing toleration.

heschlie replied:
I've updated to match the GKE tolerations, but I did leave this guy in https://github.com/projectcalico/calico/pull/1634/files#diff-4d27ea570c64ec5841f1db3b92a55c33R176

caseydavenport replied:
sgtm

Note that the comment is not quite right any more - still references the master.

heschlie replied:
Ah right, this is to enable us to run on any Node, will update the comment.

klausenbusk replied:
Do we need all this tolerations?
The following should be enough according to the [doc](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/):
> An empty key with operator Exists matches all keys, values and effects which means this will tolerate everything.
```
tolerations:
- operator: "Exists"
```

tmjd commented on master/getting-started/kubernetes/installation/hosted/kubeadm/1.7/calico.yaml:
```diff
@@ -171,17 +171,20 @@ spec:
-        operator: Exists
+        # this taint is set by all kubelets running `--cloud-provider=external`
+        # so we should tolerate it to schedule the calico pods
+        - key: node.cloudprovider.kubernetes.io/uninitialized
```
I don't think the `key: node.cloudprovider.kubernetes.io/uninitialized` one is needed anymore with the one below with no key specified. Right?

tmjd commented on master/getting-started/kubernetes/installation/hosted/kubernetes-datastore/calico-networking/1.7/calico.yaml:
```diff
@@ -102,9 +102,14 @@ spec:
+       tolerations:
+        # Allow the pod to run on the master.  This is required for
+        # the master to communicate with pods.
+        - key: node-role.kubernetes.io/master
```
Shouldn't this one be changed to `effect: NoSchedule` and `operator: exists` and the `key` dropped?

heschlie replied:
Hm, I may have been overzealous with my replacing of tolerations, I'm not sure we actually need to update Typha tolerations, I think it just needs the `CriticalAddonsOnly`

tmjd commented on master/getting-started/kubernetes/installation/hosted/kubernetes-datastore/policy-only/1.7/calico.yaml:
```diff
@@ -102,9 +102,14 @@ spec:
+       tolerations:
+        # Allow the pod to run on the master.  This is required for
+        # the master to communicate with pods.
+        - key: node-role.kubernetes.io/master
```
Same as above: Shouldn't this one be changed to `effect: NoSchedule` and `operator: exists` and the `key` dropped?

tmjd commented on master/getting-started/kubernetes/installation/hosted/calico.yaml:
```diff
@@ -106,10 +106,16 @@ spec:
-          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
-           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
     spec:
+       tolerations:
```
This same change needs to be made below to the `calico-kube-controllers`, it still has the old toleration style.

heschlie replied:
I think with the calico-kube-controllers we only need the `CriticalAddonsOnly` toleration right? If a node is flagged to keep workloads off of it (e.g. master or special nodes) I don't think we want to force kube-controllers onto it?

tmjd replied:
I don't know, we could maybe drop the NoExecute, but the previous did have a NoSchedule bit so that should maybe stay. @caseydavenport WDYT?

caseydavenport replied:
kube-controllers shouldn't tolerate _all_ taints, but it should be able to run on the master, and it should be marked as critical.

Unlike `calico/node`, I don't think it should tolerate running on any `NoSchedule` / `NoExecute` nodes.

heschlie replied:
I've included

```yaml
- key: node-role.kubernetes.io/master
          effect: NoSchedule
```

in the controllers part of the manifest tolerations.

heschlie commented:
@caseydavenport @tmjd I think this should be good to go now, I've cleaned up the `spec:` block around where I've updated/added the `tolderations:` so it is more consistent across manifests.

klausenbusk commented:
Do we need all this tolerations?
The following should be enough according to the [doc](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/):
> An empty key with operator Exists matches all keys, values and effects which means this will tolerate everything.
```
tolerations:
- operator: "Exists"
```
